async-raft/src/core/admin.rs

```diff
@@ -181,16 +181,58 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
     #[tracing::instrument(level = "trace", skip(self))]
     pub(super) async fn handle_joint_consensus_committed(&mut self) -> Result<(), RaftError> {
         if let ConsensusState::Joint { is_committed, .. } = &mut self.consensus_state {
-            *is_committed = true; // Mark as comitted.
+            *is_committed = true; // Mark as committed.
         }
         // Only proceed to finalize this joint consensus if there are no remaining nodes being synced.
         if self.consensus_state.is_joint_consensus_safe_to_finalize() {
+            self.update_replication_state().await?;
             self.finalize_joint_consensus().await?;
         }
         Ok(())
     }
 
-    /// Finalize the comitted joint consensus.
+    /// When the joint membership is committed(not the uniform membership),
+    /// a new added node turns from a NonVoter to a Follower.
+    /// Thus we need to move replication state from `non_voters` to `nodes`.
+    ///
+    /// There are two place in this code base where `nodes` are changed:
+    /// - When a leader is established it adds all node_id found in `membership` to `nodes`.
+    /// - When membership change is committed, i.e., a joint membership or a uniform membership.
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(super) async fn update_replication_state(&mut self) -> Result<(), RaftError> {
+
+        let new_node_ids = self
+            .core
+            .membership
+            .all_nodes()
+            .into_iter()
+            .filter(|elem| elem != &self.core.id)
+            .collect::<HashSet<_>>();
+
+        let old_node_ids = self.core.membership.members.clone();
+        let node_ids_to_add = new_node_ids.difference(&old_node_ids);
```

Comment:
These variable names seem a bit misleading to me, as I first expected `new_node_ids` to immediately contain solely the newly added nodes.

What do you think about something like

  * `nodes_in_new_consensus`,
  * `nodes_in_old_consensus`

?

Reply:
Great point! I had been hesitant with the names too. But seems like I forgot to fix them 😢 

`consensus` to me is too general. `node_ids_in_new_config` or `new_config_ids` feels intuitive to me.
I am quite not sure about this. English is not my native language`:(((`.
If `consensus` is more comfortable for you, I'd like to follow your advice`:DDD`.